src/components/pluginManager.js

```diff
@@ -49,103 +50,106 @@ define(['events', 'globalize'], function (events, globalize) {
-        }
-    };
-
-    // In lieu of automatic discovery, plugins will register dynamic objects
-    // Each object will have the following properties:
-    // name
-    // type (skin, screensaver, etc)
-    PluginManager.prototype.register = function (obj) {
-        this.pluginsList.push(obj);
-        events.trigger(this, 'registered', [obj]);
-    };
-
-    PluginManager.prototype.ofType = function (type) {
-        return this.pluginsList.filter(function (o) {
-            return o.type === type;
-        });
-    };
-
-    PluginManager.prototype.plugins = function () {
-        return this.pluginsList;
-    };
-
-    PluginManager.prototype.mapRoute = function (plugin, route) {
-        if (typeof plugin === 'string') {
-            plugin = this.pluginsList.filter(function (p) {
-                return (p.id || p.packageName) === plugin;
-            })[0];
+            } else if (pluginSpec.then) {
+                return pluginSpec.then(pluginBuilder => {
+                    return pluginBuilder();
+                }).then((plugin) => {
+                    console.debug(`Plugin loaded: ${plugin.id}`);
+                    return this.#registerPlugin(plugin);
+                });
+            } else {
+                const err = new TypeError('Plugins have to be a Promise that resolves to a plugin builder function or a RequireJS url (deprecated)');
+                console.error(err);
+                return Promise.reject(err);
+            }
         }
 
-        route = route.path || route;
+        // In lieu of automatic discovery, plugins will register dynamic objects
+        // Each object will have the following properties:
+        // name
+        // type (skin, screensaver, etc)
+        #register(obj) {
+            this.pluginsList.push(obj);
+            events.trigger(this, 'registered', [obj]);
+        }
 
-        if (route.toLowerCase().indexOf('http') === 0) {
-            return route;
+        ofType(type) {
+            return this.pluginsList.filter((o) => {
+                return o.type === type;
+            });
         }
 
-        return '/plugins/' + plugin.id + '/' + route;
-    };
+        #mapRoute(plugin, route) {
+            if (typeof plugin === 'string') {
+                plugin = this.pluginsList.filter((p) => {
+                    return (p.id || p.packageName) === plugin;
+                })[0];
+            }
+
+            route = route.path || route;
+
+            if (route.toLowerCase().startsWith('http')) {
+                return route;
+            }
 
-    PluginManager.prototype.mapPath = function (plugin, path, addCacheParam) {
-        if (typeof plugin === 'string') {
-            plugin = this.pluginsList.filter(function (p) {
-                return (p.id || p.packageName) === plugin;
-            })[0];
+            return '/plugins/' + plugin.id + '/' + route;
         }
 
-        var url = plugin.baseUrl + '/' + path;
+        #mapPath(plugin, path, addCacheParam) {
```

Comment:
This method is used here https://github.com/jellyfin/jellyfin-web/blob/e2aff662038bdf0fb730f9811975a02bb13a52be/src/plugins/logoScreensaver/plugin.js#L130